### PR TITLE
SliceCPU kernel to run plain memcpy when applicable

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed to improve the performance of SliceCPU kernel when the arguments are such that the slicing operation is equivalent to a copy operation.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Added special handling in SliceCPU to detect the plain copy case and run memcpy instead of the slice kernel function*
 - Affected modules and functionalities:
     *SliceCPU kernel and all the operators using it*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *New tests added*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-2138]*

### CODE REVIEW

File: dali/kernels/slice/slice_cpu.h [hunk 1]
```diff
@@ -296,12 +296,33 @@ void SliceKernel(ExecutionEngine &exec_engine,
                  const SliceArgs<OutputType, Dims> &args,
                  int min_blk_sz = 16000,
```
Comment (mzient) [inline dali/kernels/slice/slice_cpu.h:297]:
```suggestion
                 int min_blk_sz = 16 << 10,
```
?

File: dali/kernels/slice/slice_cpu.h [hunk 2]
```diff
@@ -347,6 +368,15 @@ void SliceKernel(SequentialExecutionEngine &exec_engine,
                  const SliceArgs<OutputType, Dims> &args,
                  int /* min_blk_sz */ = -1, int /* req_nblocks */ = -1) {
   (void) exec_engine;
+
+  // If the output and input data type is the same and the slice arguments take the whole extent
+  // of the input, then we can simply run memcpy.
+  if (CanRunPlainCopy<OutputType, InputType, Dims>(out_strides, in_strides, out_shape, in_shape,
+                                                   args)) {
```
Comment (mzient) [inline dali/kernels/slice/slice_cpu.h:375]: Nitpick: this line break looks weird... maybe moving all arguments to a new line with less indent or splitting between strides and shapes would look better?
```suggestion
  if (CanRunPlainCopy<OutputType, InputType, Dims>(out_strides, in_strides,
                                                   out_shape, in_shape,
                                                   args)) {
```
or
```suggestion
  if (CanRunPlainCopy<OutputType, InputType, Dims>(
      out_strides, in_strides, out_shape, in_shape, args)) {
```
?

File: dali/test/python/test_operator_slice.py [hunk 3]
```diff
@@ -678,3 +675,29 @@ def get_data():
 def test_slice_named_args_errors():
     yield check_slice_named_args_errors, 'cpu', 1
     yield check_slice_named_args_errors, 'gpu', 1
+
+def check_plain_memcpy(device, dtype, batch_size, num_threads):
```
Comment (mzient) [inline dali/test/python/test_operator_slice.py:679]: I find this test redundant. Isn't this case already covered? It's not like we can check that we're actually running plain memcpy here - and the separate test suggests just that. My recommendation is to remove it if existing tests cover that code path. Also, please verify manually (with debugger or a debug print) that the memcpy code path is actually selected.

File: dali/operators/reader/numpy_reader_op.cc [hunk 4]
```diff
@@ -240,7 +240,7 @@ void NumpyReaderCPU::RunImpl(HostWorkspace &ws) {
 
   // From 1 to 10 blocks per sample depending on the nthreads/nsamples ratio
   int blocks_per_sample = std::max(1, 10 * nthreads / nsamples);
-  constexpr int kThreshold = 16000;  // smaller samples will not be subdivided
+  constexpr int kThreshold = 16 << 10;  // smaller samples will not be subdivided
```
Comment (banasraf) [inline dali/operators/reader/numpy_reader_op.cc:243]: This constant appears in a few places. Does it make sense to name it and put it in some common place?

File: dali/kernels/slice/slice_kernel_utils.h [hunk 5]
```diff
@@ -104,6 +104,31 @@ inline std::tuple<int64_t, int64_t, int64_t> CalcPadCopyExtents(int64_t anchor,
...
+  // if there's type conversion we can't run memcpy
+  if (!std::is_same<OutputType, InputType>::value)
+    return false;
+
+  auto default_out_strides = GetStrides(out_shape);
+  auto default_in_strides = GetStrides(in_shape);
+
+  // If the strides are not the default ones, or the window anchor and shape
+  // are different than the bounds of the input, we can't run plain memcpy
+  for (int d = 0; d < Dims; d++) {
```
Comment (banasraf) [inline dali/kernels/slice/slice_kernel_utils.h:123]: Isn't there a special case for 1D data that you could use memcpy even when the anchor and shape are different than the input bounds, but assuming the default stride?

Comment (banasraf) [reply dali/kernels/slice/slice_kernel_utils.h:123]: I mean, it still would be a contiguous chunk of memory.

Comment (jantonguirao) [reply dali/kernels/slice/slice_kernel_utils.h:123]: Yeah, I thought about that, but decided not to over-engineer it too much

File: dali/operators/reader/numpy_reader_op.cc (hunk 4 above)
Comment (jantonguirao) [reply dali/operators/reader/numpy_reader_op.cc:243]: Makes sense. Will do

**Review by banasraf — APPROVED** [review]

**Review by szalpal — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**